### PR TITLE
 Download data dictionary as csv or xlsx #178 

### LINKED ISSUE #178: Download data dictionary as csv or xlsx
**OP timer**

https://openplus.monday.com/boards/4092908516/pulses/5233179959

---

## User story

As a data catalogue user, I want to download the data dictionary so I can work with it in my preferred tools.
As a data catalogue editor, I want to be able to download the existing data dictionary so that I can make changes in the software of my choice, then upload the changes using the existing import feature.

## This feature support the following requirements

[alpha release](https://openplus.monday.com/boards/3958997205/pulses/4605939234)

## Additional context

This feature should build on features that exist already. We have features already that can import and export csv and excel. 

-  #22  
- #115 

The data dictionary is in the entity reference to a paragraph , see `field_columns` on content type `data_set`. The label for dataset is _Metadata record_

## Proposed solution

- Developer to determine the details.
- On the import columns build page, `/node/nid/add-columns`, there should be an option to download the data dictionary as csv. This should be conditional, if there are no data columns, there should be no download option.
- On the node view, there are links to download either a csv or and excel copy of the data dictionary.  This should be conditional, if there are no data columns, there should be no download option.

### Screenshots

Download links as they display on the node view

Download links as they appear on the build/add-columns page

## Definition of done (DoD)
- [x] Anonymous users can download a data dictionary from a metadata record in csv format
- [x] Anonymous users can download a data dictionary from a metadata record in xlsx format
- [x] Anonymous users cannot see any download links for the data dictionary if the dictionary does not exist
- [x] Data catalogue editors can download the existing data catalogue as csv on `/node/nid/add-columns`
- [x] Data catalogue editors cannot see a download link if there is no data dictionary
- [x] the data dictionary that is downloaded can be imported back into the app using the upload feature on `/node/nid/add-columns`

### Automated functional tests

- [x] I have written functional tests for this feature
- [x] I have run the functional tests and they pass

### PR DESCRIPTION
**Pull request by jhoude5**
#178

### CODE REVIEW

**Review by joel-osc — COMMENTED** [review]
Just a few comments:

1. I think Excel should be xlsx which is the actual format, not the name of the software app
2. Seeing some drupal code standard things with position of { } and if/for - maybe the standards have changed and I am not aware but there are definitely some inconsitancies. 
3. If you are using getTranslationFromContext then shouldn't the column titles also be wrapped in "t"?

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+use vendor\PhpOffice\PhpSpreadsheet\Reader\Xlsx;
+
+class BcDcCreateFileController extends ControllerBase {
+
+  /**
+   * Returns a downloadable file.
+   */
+  public function createFile($node, $param) {
+  
+    if ($param === 'csv' || $param === 'xlsx') {
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: Have an array of allowed types. If `$param` is not in the array, throw the exception. It is difficult to understand an `if` statement that is so long. It is easier to read to have short conditions near the start that return to handle the error states and then have the body of the function not inside any `if` statements.

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+          $paragraph->get('field_provenance_field_number')->value,
+          $paragraph->get('field_provenance_field_question')->value,
+          $paragraph->get('field_provenance_form_name')->value,
+          $paragraph->get('field_provenance_form_number')->value,
+          ]
+      ];
+
+      $spreadsheet = new Spreadsheet();
+
+      if ($param === 'xlsx') {
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: Use a `switch` statement for this.

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        $writer->save($directory);
+
+      }
+
+    }
+    else {
+      throw new NotFoundHttpException();
+    }
+
+    $uri = $directory;
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: Why is there both `$uri` and `$directory`? They seem to be the same thing.

File: html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php [hunk 4]
```diff
@@ -544,6 +545,14 @@ public function test(): void {
     $this->assertSession()->statusCodeEquals(200);
     // Page has ISO dates.
     $this->isoDateTest();
+    
+    // Anonymous has access to download csv of data dictionary.
+    $this->drupalGet('node/19/download/columns/csv');
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php]: This looks like the old test that didn't pass.

File: html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php [hunk 5]
```diff
@@ -97,6 +97,7 @@ protected function createTestUser(string $name, array $roles = []): User|false {
    * Tests.
    */
   public function test(): void {
+
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php]: Why add this space?

File: html/themes/custom/dc_theme/templates/node/node--data-set--full.html.twig [hunk 6]
```diff
@@ -194,6 +194,10 @@
       {% if node.field_modified_date and not node.field_modified_date.isempty %}
       <hr>
       <h2>{{ 'Data dictionary'|t }}</h2>
+        {% if node.field_columns and not node.field_columns.isempty %}
+          <p><a href="/node/{{ node.id }}/download/columns/csv">Export data dictionary to csv</a></p>
```
Comment (lkmorlan) [inline html/themes/custom/dc_theme/templates/node/node--data-set--full.html.twig]: Each download link should be a `ul > li`.
Use "CSV" and "XLSX". Better to use the name of the format than the application.
It would be better to generate this list of links in template pre-process and just output the variable here. The variable would just be NULL if there are no columns.
There should be a sngle function the also generates the list for `bc_dc_form_alter()`.

File: html/modules/custom/bc_dc/bc_dc.module [hunk 7]
```diff
@@ -276,3 +276,22 @@ function bc_dc_cron(): void {
   $reviewReminder = \Drupal::service('bc_dc.review_reminder');
   $reviewReminder->sendRemindersToAllUsers();
 }
+
+/**
+ * Implements hook_form_alter.
+ */
+function bc_dc_form_alter(&$form, \Drupal\Core\Form\FormStateInterface $form_state, $form_id) {
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/bc_dc.module]: Add a full set of type declarations.

File: html/modules/custom/bc_dc/bc_dc.module [hunk 8]
```diff
@@ -276,3 +276,22 @@ function bc_dc_cron(): void {
...
+  if ($form['form_id']['#id'] === 'edit-bc-dc-add-columns-form') {
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/bc_dc.module]: Use `hook_form_FORM_ID_alter()` so this line isn't needed.

File: html/modules/custom/bc_dc/bc_dc.module [hunk 9]
```diff
@@ -276,3 +276,22 @@ function bc_dc_cron(): void {
...
+    $nid = $form_state->getStorage('nid');
+    $paragraph = \Drupal::entityTypeManager()->getStorage('node')->load($nid['nid']);
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/bc_dc.module]: Check that the node is the correct bundle.
`$paragraph` should be `$entity`.

File: html/modules/custom/bc_dc/bc_dc.module [hunk 10]
```diff
@@ -276,3 +276,22 @@ function bc_dc_cron(): void {
...
+    $paragraph_field_items = $paragraph->get('field_columns')->referencedEntities();
+    if ($paragraph_field_items) {
+      $form['data_dictionary'] = [
+        '#markup' =>
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/bc_dc.module]: This should use `item_list`.

File: html/modules/custom/bc_dc/bc_dc.module [hunk 11]
```diff
@@ -276,3 +276,22 @@ function bc_dc_cron(): void {
...
+          '<p><a href="/node/' . $nid['nid'] . '/download/columns/csv">Download current data dictionary file to CSV</a></p>
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/bc_dc.module]: These should be generated with `Link::createFromRoute()`.

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            $paragraph->get('field_provenance_field_number')->value,
+            $paragraph->get('field_provenance_field_question')->value,
+            $paragraph->get('field_provenance_form_name')->value,
+            $paragraph->get('field_provenance_form_number')->value,
+            ]
+        ];
+
+        $spreadsheet = new Spreadsheet();
+
+        if ($param === 'xlsx') {
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: Use a `switch` statement for this.

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+            $writer->setEnclosure('');
+            $writer->setLineEnding("\r\n");
+            $writer->setSheetIndex(0);
+            $writer->save($directory);
+        }
+    } else {
+        throw new NotFoundHttpException();
+    }
+
+        $uri = $directory;
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: Why does `$directory` exist?

**Review by lkmorlan — CHANGES_REQUESTED** [review]

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+   * Returns a downloadable file.
+   */
+public function createFile($node, $param) 
+{
+    $nid = $node->get('nid')->getValue()[0]['value'];
+    $alias = \Drupal::service('path_alias.manager')->getAliasByPath('/node/'.$nid);
+    $path = str_replace("/data-set/", "", $alias);
+
+    if ($param === 'csv' || $param === 'xlsx') {
+        $paragraph = \Drupal::entityTypeManager()->getStorage('node')->load($nid);
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: Use `$this->entityTypeManager()`.

**Review by lkmorlan — CHANGES_REQUESTED** [review]

File: html/modules/custom/bc_dc/bc_dc.module (hunk 10 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/bc_dc.module]: @lkmorlan which line should be 'item_list'?

File: html/modules/custom/bc_dc/bc_dc.module (hunk 9 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/bc_dc.module]: fixed

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 3 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: fixed

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 2 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: fixed

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 1 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: fixed

File: html/modules/custom/bc_dc/bc_dc.module (hunk 10 above)
Comment (lkmorlan) [reply html/modules/custom/bc_dc/bc_dc.module]: I mean to theme it as `'#type' => 'item_list'` instead of using `#markup`.

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 14 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: fixed

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 13 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: removed

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 12 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: fixed

File: html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php (hunk 5 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php]: fixed

File: html/modules/custom/bc_dc/bc_dc.module (hunk 10 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/bc_dc.module]: fixed

File: html/modules/custom/bc_dc/bc_dc.module (hunk 11 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/bc_dc.module]: fixed

**Review by lkmorlan — DISMISSED** [review]
I suggest code like this:

```
$formats = [
	'csv',
	'xlsx',
];
$items = [];
foreach ($formats as $format) {
	$items[] = Link::createFromRoute(
		t('Download current data dictionary file to @format', ['@format'] => strtoupper($format)),
		'bc_dc.create_columns_file',
		['node' => $nid['nid'], 'param' => $format],
	);
}
```

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 15]
```diff
@@ -4,105 +4,105 @@
...
+          $directory = 'public://data_dictionary/' . $filename;
+          $writer->setDelimiter(',');
+          $writer->setEnclosure('');
+          $writer->setLineEnding("\r\n");
+          $writer->setSheetIndex(0);
+          $writer->save($directory);
+      }
+    }
+    else {
+      throw new NotFoundHttpException();
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php:71]: Move this to near the top of the function:

```
if (!in_array($param, $allowedExt, TRUE)) {
  throw new NotFoundHttpException();
}
```
The rest of the function would not be inside the `if` statement at all.

Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php:71]: fixed

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+    // Generate the spreadsheet object and save to $file_path.
+    $spreadsheet = new Spreadsheet();
+    $worksheets = new Worksheet($spreadsheet);
+    $spreadsheet->addSheet($worksheets);
+    $worksheets->fromArray($sheetData);
+    $sheetIndex = $spreadsheet->getIndex(
+      $spreadsheet->getSheetByName('Worksheet')
+    );
+    $spreadsheet->removeSheetByIndex($sheetIndex);
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php:105]: Why is this needed? Does a new spreadsheet come with a blank worksheet that needs to be removed? If so, a comment to this effect would be good.

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+    foreach ($worksheets as $worksheet) {
+      foreach ($worksheet->getColumnIterator() as $column) {
+        $worksheet->getColumnDimension($column->getColumnIndex())->setAutoSize(TRUE);
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: A comment about what this is for would be welcome.

File: html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php [hunk 18]
```diff
@@ -536,6 +536,13 @@ public function test(): void {
     // Page has ISO dates.
     $this->isoDateTest();
 
+    // Anonymous has access to download csv for Metadata record when file has been uploaded
+    $this->drupalGet('node/2/download/columns/csv');
+    $this->assertSession()->statusCodeEquals(200);
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php:542]: It would be good to test that the contents of this file are as expected.

**Review by lkmorlan — CHANGES_REQUESTED** [review]

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+        'column_data_quality',
+        'column_description',
+        'column_name',
+        'column_size',
+        'column_transformations',
+        'provenance_field_name',
+        'provenance_field_number',
+        'provenance_field_question',
+        'provenance_form_name',
+        'provenance_form_number',
```
Comment (lkmorlan) [inline html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: It would be best if it did not have a hard-coded list of fields and instead included every field in the paragraph type. You can refer to the code that generates the "Download sample file" for how to do this.

**Review by lkmorlan — CHANGES_REQUESTED** [review]

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 16 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php:105]: Comment had been added

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 17 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: Removed as this is not really needed

File: html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php (hunk 18 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php:542]: Not sure how to test this

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 19 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: Not sure from the code in Download sample will fix hardcoding the headers

Comment (lkmorlan) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: The download sample link has code to get the names of all the fields so it does not have to have its own list.

File: html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php (hunk 18 above)
Comment (lkmorlan) [reply html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php:542]: You can probably use `::assertSession()->responseContains()`.

Comment (jhoude5) [reply html/modules/custom/bc_dc/tests/src/Functional/BcDcFunctionalTest.php:542]: Used this assert

File: html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php (hunk 19 above)
Comment (jhoude5) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: updated code to not hardcode headers

Comment (lkmorlan) [reply html/modules/custom/bc_dc/src/Controller/BcDcCreateFileController.php]: The rows are still generated with a hard-coded list of fields.

**Review by lkmorlan — APPROVED** [review]
